proposals/0046-i18n-domain-support.md

```diff
@@ -0,0 +1,87 @@
+- Start Date: 2023/12/21
+- Reference Issues:
+- Implementation PR: https://github.com/withastro/astro/pull/9143
+
+
+# Summary
+
+First-class support for domain support in i18n routing.
+
+
+# Background & Motivation
+
+Websites that support internationalisation have different requirements. Among these requirements, there's the need to 
+have localised content under different subdomains or domains. 
+
+# Goals
+
+- Support different domains, all driven by the same Astro project;
+- Configure some locales to use only path prefixes and some locales to use domains;
+- Mix locales that require a different domain, with locales that don't require a different domain;
+
+# Non-Goals
+
+- Redirect users from a path website to a domain website 
+- Change website/domain based on the language of the user's browser 
+- Support for static/hybrid output
+- Offer the means to configure multiple domains for serverless hosts (Vercel, Netlify, etc.)
+
+
+# Detailed Design
+
+There are some restrictions about how the feature will initially work:
+- the feature works only with `output: "server"`, hence only in SSR;
+- the presence of at least one pre-rendered route will cause a failure in the build;
+- the option `base` is required, because Astro needs to know how to create absolute URLs for locales that aren't mapped to a domain;
+- `functionPerRoute` isn't supported
+
+A feature that allows to support different domains for certain locales.
+
+Using the configuration `domains`, a user can specify which locales should benefit from a domain. This feature changes the behaviour of some of the APIs exported by the virtual module `astro:i18n`.
+
+```js
+// astro.config.mjs
+import {defineConfig} from "astro/config"
+export default defineConfig({
+    i18n: {
+        defaultLocaLe: 'en',
+        locales: ['en', 'es', 'pt_BR', 'pt', 'fr'],
+        domains: {
+            fr: "https://fr.example.com",
+            pt: "https://example.pt"
+        },
+        routingStrategy: "domain"
+    }
+})
+```
+
+The following APIs will behave as follows:
+- [`getRelativeLocaleUrl`](#getrelativelocaleurllocale-string-string): it won't prefix the locale to the URL. From `/en` to `/`;
```

Comment:
Wondering if this should even be an error or something in this case? Seems like kind of a footgun if you use `getRelativeLocaleUrl()` hoping to navigate to a different locale, but get a same-domain random path instead.

Also to make sure I follow this. With the example config above, would you get the following behaviour?

```js
getRelativeLocaleUrl('en', '') // => '/en'
getRelativeLocaleUrl('es', '') // => '/es'
getRelativeLocaleUrl('fr', '') // => '/'
```

And is that consistent across domains? i.e. for a page on `fr.example.com`, does `getRelativeLocaleUrl('en', '')` still return `'/en'`? That would resolve to `fr.example.com/en`, which seems kind of useless.

Should sites using multiple domains perhaps _always_ use `getAbsoluteLocaleUrl()`? At the most extreme end, we could even change the type of `astro:i18n` and remove `getRelativeLocaleUrl()` entirely in these cases.

Reply:
> And is that consistent across domains? i.e. for a page on fr.example.com, does getRelativeLocaleUrl('en', '') still return '/en'? That would resolve to fr.example.com/en, which seems kind of useless.

No, it's not consistent and it's impossible to make it consistent. 

> Should sites using multiple domains perhaps always use getAbsoluteLocaleUrl()? At the most extreme end, we could even change the type of astro:i18n and remove getRelativeLocaleUrl() entirely in these cases.

Is it something that we can do? 😮 If so, I think we should go this way, although it should make the transition from a non-domain website to a domain website.

Maybe we should do something in the middle; if the user uses a `getRelativeLocale*` function, we should fall back to use `getAbsoluteLocale*` instead. And we could add a warning.

Reply:
Also, I realised that `getRelativeLocaleUrl` should not change its behaviour. I removed that line

Reply:
> Maybe we should do something in the middle; if the user uses a `getRelativeLocale*` function, we should fall back to use `getAbsoluteLocale*` instead. And we could add a warning.

Yes, that would be a nice compromise I think and wouldn’t need the type to change (I assume? The two functions have the same signature right?) while still giving feedback that the function used is not intended for this context.